rzup/src/utils.rs

```diff
@@ -0,0 +1,184 @@
+        Ok(())
+    }
+
+    fn output_if_success(&mut self) -> Result<Output> {
+        let cmd = self.as_command_mut();
+        let output = cmd
+            .output()
+            .with_context(|| format!("failed to create process {:?}", cmd))?;
+        check_success(cmd, &output.status, &output.stdout, &output.stderr)?;
+        Ok(output)
+    }
+}
+
+impl CommandExt for Command {
+    fn as_command_mut(&mut self) -> &mut Command {
+        self
+    }
+}
+
+pub fn check_success(
+    cmd: &Command,
+    status: &ExitStatus,
+    stdout: &[u8],
+    stderr: &[u8],
+) -> Result<()> {
+    if status.success() {
+        return Ok(());
+    }
+    Err(ProcessError {
+        cmd_desc: format!("{:?}", cmd),
+        status: *status,
+        stdout: stdout.to_vec(),
+        stderr: stderr.to_vec(),
+        hidden: false,
+    }
+    .into())
+}
+
+#[derive(Debug)]
+struct ProcessError {
+    status: ExitStatus,
+    #[allow(dead_code)]
+    hidden: bool,
+    stdout: Vec<u8>,
+    stderr: Vec<u8>,
+    cmd_desc: String,
+}
+
+impl fmt::Display for ProcessError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "failed to execute {}", self.cmd_desc)?;
+        write!(f, "\n    status: {}", self.status)?;
+        if !self.stdout.is_empty() {
+            let stdout = String::from_utf8_lossy(&self.stdout);
+            let stdout = stdout.replace('\n', "\n        ");
+            write!(f, "\n    stdout:\n        {}", stdout)?;
+        }
+        if !self.stderr.is_empty() {
+            let stderr = String::from_utf8_lossy(&self.stderr);
+            let stderr = stderr.replace('\n', "\n        ");
+            write!(f, "\n    stderr:\n        {}", stderr)?;
+        }
+        Ok(())
+    }
+}
+
+impl std::error::Error for ProcessError {}
+
+pub struct FileLock(File);
+
+impl Drop for FileLock {
+    fn drop(&mut self) {
+        drop(self.0.unlock());
+    }
+}
+
+pub fn flock(path: &Path) -> Result<FileLock> {
+    let parent = path.parent().unwrap();
+    std::fs::create_dir_all(parent)
+        .context(format!("failed to create directory `{}`", parent.display()))?;
+    let file = OpenOptions::new()
+        .create(true)
+        .read(true)
+        .write(true)
+        .open(path)?;
+    file.lock_exclusive()?;
+    return Ok(FileLock(file));
+}
+
+/// Try to get the host target triple.
+///
+/// Only checks for targets that have pre-built toolchains.
+#[allow(unreachable_code)]
+pub fn guess_host_target() -> Option<&'static str> {
+    #[cfg(all(target_arch = "x86_64", target_os = "linux"))]
+    return Some("x86_64-unknown-linux-gnu");
+
+    #[cfg(all(target_arch = "x86_64", target_os = "macos"))]
+    return Some("x86_64-apple-darwin");
+
+    #[cfg(all(target_arch = "aarch64", target_os = "macos"))]
+    return Some("aarch64-apple-darwin");
+
+    #[cfg(all(target_arch = "x86_64", target_os = "windows"))]
+    return Some("x86_64-pc-windows-msvc");
+
+    None
+}
```

Comment:
Using `cfg_if` might be good here, and maybe this could be a `const HOST_TARGET_TRIPLE: &str = ...`

Reply:
I think I borrowed this code from some other codebase (which is why I'm to blame for this originally)